File: logitdyn.py
"""
Filename: logitdyn.py

Authors: Daisuke Oyama

Logit dynamics.

"""
import numpy as np
import numbers
from util import check_random_state
from normal_form_game import *
from random_game import random_pure_actions


class LogitDynamics(object):
    """
    Parameters
    ----------
    g : NormalFormGame
        N-player game played

    beta : scalar(float)

    Attributes
    ----------
    N : scalar(int)
        The number of players in the game

    players : list(Player)
        The list consisting of all players with the given payoff matrix.

    nums_actions : tuple(int)
        Tuple of the number of actions, one for each player.

    beta : scalar(float)
        See Parameters.

    player.logit_choice_cdfs : array-like
        The choice probability of each actions given opponents' actions.
    """
    def __init__(self, g, beta=1.0):
        self.N = g.N
        self.players = g.players
        self.nums_actions = g.nums_actions

        self.beta = beta

        for player in self.players:
            payoff_array_rotated = \
                player.payoff_array.transpose(list(range(1, self.N)) + [0])
            # Shift payoffs so that max = 0 for each opponent action profile
            payoff_array_rotated -= \
                payoff_array_rotated.max(axis=-1)[..., np.newaxis]
            # cdfs left unnormalized
            player.logit_choice_cdfs = \
                np.exp(payoff_array_rotated*self.beta).cumsum(axis=-1)
            # player.logit_choice_cdfs /= player.logit_choice_cdfs[..., [-1]]

    def logit_choice_cdfs(self):
        return tuple(player.logit_choice_cdfs for player in self.players)

    def _play(self, player_ind, actions, random_state=None):
        random_state = check_random_state(random_state)
        i = player_ind

        # Tuple of the actions of opponent players i+1, ..., N, 0, ..., i-1
        opponent_actions = \
            tuple(actions[i+1:]) + tuple(actions[:i])

        cdf = self.players[i].logit_choice_cdfs[opponent_actions]
        random_value = random_state.rand()
        next_action = cdf.searchsorted(random_value*cdf[-1], side='right')

        return next_action

    def play(self, init_actions=None, player_ind_seq=None, num_reps=1,
             random_state=None):
        random_state = check_random_state(random_state)
        if init_actions is None:
            init_actions = random_pure_actions(self.nums_actions, random_state)
        init_actions = list(init_actions)

        if player_ind_seq is None:
            player_ind_seq = random_state.randint(self.N, size=num_reps)

        if isinstance(player_ind_seq, numbers.Integral):
            player_ind_seq = [player_ind_seq]

        for t, player_ind in enumerate(player_ind_seq):
            init_actions[player_ind] = self._play(player_ind, init_actions,
                                                  random_state)

        return tuple(init_actions)

    def time_series(self, ts_length, init_actions=None, random_state=None):
        random_state = check_random_state(random_state)
        if init_actions is None:
            init_actions = random_pure_actions(self.nums_actions, random_state)
        actions = [action for action in init_actions]

        out = np.empty((ts_length, self.N), dtype=int)
        player_ind_seq = random_state.randint(self.N, size=ts_length)

        for t, player_ind in enumerate(player_ind_seq):
            for i in range(self.N):
                out[t, i] = actions[i]
            actions[player_ind] = self._play(player_ind, actions, random_state)

        return out

File: logitdyn.jl
#=
    Tools for Logit Response Dynamics
=#

using Games
using Random


# LogitDynamics #

"""
    LogitDynamics{N, T, S}

Type representing Logit-Dynamics model.

# Fields

- `players::NTuple{N,Player{N,T}}` : Tuple of player instances.
- `nums_actions::NTuple{N,Int}` : Tuple of integers which are the number of
    actions for each player.
- `beta::S` : The level of noise in players' decision.
- `choice_probs` : Each players' choice probabilities for each actions.
"""
struct LogitDynamics{N,T<:Real,S<:Real}
    players::NTuple{N,Player{N,T}}
    nums_actions::NTuple{N,Int}
    beta::S
    choice_probs
end

"""
    LogitDynamics(g, beta)

Construct a `LogitDynamics` instance.

# Arguments

- `g::NormalFormGame{N,T}` : `NormalFormGame` instance.
- `beta::S` : The level of noise in players' decision.

# Return

- `::LogitDynamics` : New `LogitDynamics` instance.
"""
function LogitDynamics(g::NormalFormGame{N,T}, beta::S) where {N,T<:Real,S<:Real}
    choice_probs = Vector{Any}(undef, N)
    for (i, player) in enumerate(g.players)
        payoff_array = player.payoff_array
        payoff_array_normalized = payoff_array .- maximum(payoff_array, dims=N)
        choice_probs[i] = cumsum(exp.(payoff_array_normalized .* beta),
                                       dims=N)
    end
    return LogitDynamics(g.players, g.nums_actions, beta, choice_probs)
end

"""
    play!([rng,] lgdy, player_ind, actions)

Return new action of player indexed by `player_ind` given each players' choice
probabilities.

# Arguments

- `rng::AbstractRNG` : Random number generator used.
- `lgdy::LogitDynamics{N}` : `LogitDynamics` instance.
- `player_ind::Integer` : Player index who takes action.
- `actions::Vector{<:Integer}` : Vector of actions for each players.

# Return

- `::Integer` : The new action of player indexed by `player_ind`.
"""
function play!(rng::AbstractRNG, lgdy::LogitDynamics{N}, player_ind::Integer,
               actions::Vector{<:Integer}) where N
    oppponent_actions = [actions[player_ind+1:N]..., actions[1:player_ind-1]...]
    cdf = lgdy.choice_probs[player_ind][oppponent_actions..., :]
    random_value = rand(rng)
    next_action = searchsortedfirst(cdf, random_value*cdf[end])
    return next_action
end

"""
    play([rng,] lgdy, init_actions; num_reps)

Return new action profile after `num_reps` iterations.

# Arguments

- `rng::AbstractRNG` : Random number generator used.
- `lgdy::LogitDynamics{N}` : `LogitDynamics` instance.
- `init_actions::Games.PureActionProfile` : Initial action profile.
- `num_reps::Integer` : The number of iterations; defaults to 1.

# Return

- `::Vector{<:Integer}` : New action profile.
"""
function play(rng::AbstractRNG,
              lgdy::LogitDynamics{N},
              init_actions::Games.PureActionProfile;
              num_reps::Integer=1) where N
    actions = [m for m in init_actions]
    player_ind_seq = rand(rng, 1:N, num_reps)
    for player_ind in player_ind_seq
        actions[player_ind] = play!(rng, lgdy, player_ind, actions)
    end
    return actions
end

play(lgdy::LogitDynamics, init_actions::Games.PureActionProfile;
     num_reps::Integer=1) =
    play(Random.GLOBAL_RNG, lgdy, init_actions, num_reps=num_reps)

"""
    time_series!(rng, lgdy, out, player_ind_seq)

Update `out` which represents the time series of action profile.

# Arguments

- `rng::AbstractRNG` : Random number generator used.
- `lgdy::LogitDynamics{N}` : `LogitDynamics` instance.
- `out::Matrix{<:Integer}` : Matrix which represents the time series of action
    profile.
- `player_ind_seq::Vector{<:Integer}` : The sequence of player index, which is
    determined randomly.

# Return

- `::Matrix{<:Integer}` : Updated `out`.
"""
function time_series!(rng::AbstractRNG,
                      lgdy::LogitDynamics{N},
                      out::Matrix{<:Integer},
                      player_ind_seq::Vector{<:Integer}) where N
    ts_length = size(out, 2)
    current_actions = [out[i, 1] for i in 1:N]
    for t in 1:ts_length-1
        current_actions[player_ind_seq[t]] = play!(rng, lgdy, player_ind_seq[t],
                                                   current_actions)
        for i in 1:N
            out[i, t+1] = current_actions[i]
        end
    end
    return out
end

"""
    time_series([rng,] lgdy, ts_length, init_actions)

Return the time series of action profile

# Arguments

- `rng::AbstractRNG` : Random number generator used.
- `lgdy::LogitDynamics{N}` : `LogitDynamics` instance.
- `ts_length::Integer` : The length of time series.
- `init_actions::Games.PureActionProfile` : Initial action profile.
"""
function time_series(rng::AbstractRNG,
                     lgdy::LogitDynamics{N},
                     ts_length::Integer,
                     init_actions::Games.PureActionProfile) where N
    player_ind_seq = rand(rng, 1:N, ts_length-1)
    out = Matrix{Int}(undef, N, ts_length)
    for i in 1:N
        out[i, 1] = init_actions[i]
    end
    time_series!(rng, lgdy, out, player_ind_seq)
end

time_series(lgdy::LogitDynamics, ts_length::Integer,
            init_actions::Games.PureActionProfile) =
    time_series(Random.GLOBAL_RNG, lgdy, ts_length, init_actions)
